model/stepfun.go

```diff
@@ -0,0 +1,253 @@
+	topP        float32
+}
+
+func NewStepFunModelProvider(subType string, apiKey string, temperature float32, topP float32) (*StepFunModelProvider, error) {
+	return &StepFunModelProvider{
+		subType:     subType,
+		apiKey:      apiKey,
+		temperature: temperature,
+		topP:        topP,
+	}, nil
+}
+
+func (p *StepFunModelProvider) GetPricing() string {
+	return `URL:
+https://platform.stepfun.com/docs/pricing/details
+
+| Model        | Input Price per 1K characters    | Output Price per 1K characters |
+|--------------|----------------------------------|--------------------------------|
+| step-1-8k    | 0.005yuan/1,000 tokens           | 0.02yuan/1,000 tokens          |
+| step-1-32k   | 0.015yuan/1,000 tokens           | 0.07yuan/1,000 tokens          |
+| step-1-128k  | 0.04yuan/1,000 tokens            | 0.2yuan/1,000 tokens           |
+| step-1-256k  | 0.095yuan/1,000 tokens           | 0.3yuan/1,000 tokens           |
+| step-1-flash | 0.001yuan/1,000 tokens           | 0.004yuan/1,000 tokens         |
+| step-2-16k   | 0.038yuan/1,000 tokens           | 0.12yuan/1,000 tokens          |
+`
+}
+
+func (p *StepFunModelProvider) calculatePrice(modelResult *ModelResult) error {
+	price := 0.0
+	priceTable := map[string][2]float64{
+		"step-1-8k":    {0.005, 0.02},
+		"step-1-32k":   {0.015, 0.07},
+		"step-1-128k":  {0.04, 0.2},
+		"step-1-256k":  {0.095, 0.3},
+		"step-1-flash": {0.001, 0.004},
+		"step-2-16k":   {0.038, 0.12},
+	}
+
+	if priceItem, ok := priceTable[p.subType]; ok {
+		inputPrice := getPrice(modelResult.PromptTokenCount, priceItem[0])
+		outputPrice := getPrice(modelResult.ResponseTokenCount, priceItem[1])
+		price = inputPrice + outputPrice
+	} else {
+		return fmt.Errorf("calculatePrice() error: unknown model type: %s", p.subType)
+	}
+
+	modelResult.TotalPrice = price
+	modelResult.Currency = "CNY"
+	return nil
+}
+
+type ChatCompletionMessage struct {
+	Role    string `json:"role"`
+	Content string `json:"content"`
+}
+
+type TokenCountRequest struct {
+	Model    string                  `json:"model"`
+	Messages []ChatCompletionMessage `json:"messages"`
+}
+
+type TokenCountResponse struct {
+	Data struct {
+		TotalTokens int `json:"total_tokens"`
+	} `json:"data"`
+}
+
+func StepFunNumTokensFromMessages(prompt string, response string, subType string, apiKey string) (int, error) {
+	user_prompt := ChatCompletionMessage{
+		Role:    "user",
+		Content: prompt,
+	}
+	sys_response := ChatCompletionMessage{
+		Role:    "system",
+		Content: response,
+	}
+	var messages TokenCountRequest
+	if response != "" {
+		messages = TokenCountRequest{
+			Model:    subType,
+			Messages: []ChatCompletionMessage{user_prompt, sys_response},
+		}
+	} else {
+		messages = TokenCountRequest{
+			Model:    subType,
+			Messages: []ChatCompletionMessage{user_prompt},
+		}
+	}
+
+	requestBodyJSON, err := json.Marshal(messages)
+	if err != nil {
+		log.Fatalf("JSON Marshal Error: %v", err)
+	}
+
+	url := "https://api.stepfun.com/v1/token/count"
```

Comment:
Why call another API to calculate tokens? Their docs (https://platform.stepfun.com/docs/pricing/intro) said, they already returned the token usage in the response?

![image](https://github.com/user-attachments/assets/cb0c553f-1e32-459d-b743-db6a255c4b76)